[글로] 2단계 코틀린 자동차 경주 제출합니다.

ippnsj opened the pull request:
1단계 피드백을 모두 수정하고 푸쉬하려다가 1단계에서 아무것도 수정하지 못하고 merge가 되어버렸네요:disappointed_relieved:
다음에는 1단계에서도 수정할 수 있도록 노력하겠습니다!

1단계 피드백을 토대로 수정해 보았습니다!
1. Controller의 cars 멤버 변수의 경우, 재활용 가능한 방법이 뭘까 고민을 많이 하다가 Cars 라는 일급 컬렉션을 만들어 보는건 어떨까 생각했습니다!
2. 입력을 다시 받는 것은 InputView의 일로 분리해 보라는 피드백을 받고 검증 위치에 대해 많이 고민했습니다. 슬랙에 질문을 올리고 여러 크루들로부터 값을 직접 사용하는 도메인에서 검증을 하는 것이 좋지 않을까요 라는 의견을 받고 도메인에서의 검증을 생각하고 있었기 때문에 고민이 컸던 것 같아요! 하지만 InputView에서 입력을 다시 받아야 하는지 여부를 결정하기 위해서는 검증이 필요하다고 생각했고, 저 또한 입력을 다시 받는 것은 컨트롤러가 아닌 InputView의 역할이 맞다고 생각했습니다!
결과적으로 도메인에서 값을 전달 받은 직후에도, InputView가 입력을 받은 직후에도 검증이 필요하다고 생각하여 두 곳에서 모두 검증을 진행하였습니다~!:smile:
3. 남겨주신 프로퍼티 가시성 문서를 보고 주생성자에서 생성된 프로퍼티의 getter를 어떻게 가시성 높게 표현할 수 있을까 고민했습니다. 그러다가 그냥 접근제한자를 public으로 바꾸면 좋지 않을까 라는 생각을 하게 되었고 그렇게 수정했습니다! 더 좋은 방법이 있다면 알려주세요~!:blush:
4. Inner class 였던 TestNumberGenerator를 테스트 코드에 익명 클래스로 넘겨주었습니다! 람다로 넘기라는 말씀에 부합하는지는 잘 모르겠습니다:sob:
5. 그동안은 하드 코딩은 무조건 나쁘다고만 생각해 보았지 하드 코딩이 더 읽기 편할수도 있다는 생각을 해보지 못 했던 것 같아요! 피드백을 받고 테스트 코드를 하드 코딩 했을 때와 안 했을 때를 비교해 보았을 때, 오히려 하드 코딩하지 않았을 때 더 복잡하다고 느꼈습니다!

피드백 내용을 보고 제가 이해한대로 수정해 보았습니다! 혹시 잘못 이해하여 수정한 부분이 있다면 알려주세요!ㅠㅠ 리뷰어님 덕분에 코드는 읽기 편해야 한다는 점을 다시 상기하게 되었고, 생각해보지 못한 부분도 생각해 볼 수 있었습니다! 감사합니다~!:heart_eyes:

ippnsj committed:
refactor: 우승자 찾기 기능 수정

가장 긴 거리를 찾고 같은 거리를 가진 우승자를 찾는 방식으로 기능 단순화

ippnsj committed:
refactor: 코틀린 문법 활용

Boolean 리턴 한 줄로 단축

ippnsj committed:
refactor: SAM 익명 클래스로 전달

ippnsj committed:
refactor: cars 일급 컬랙션 생성

cars의 재활용성을 높이기 위해 멤버 변수 대신 일급 컬랙션으로 생성
이에 따라 Judgment 클래스의 역할을 Cars 내부에서 진행
DTO에서 도메인 역할인 객체 리스트 생성을 하지 않게 됨

ippnsj committed:
refactor: dto에서 검증 로직 분리

입력별 검증 클래스 추가

ippnsj committed:
refactor: 재입력 기능을 InputView의 일로 분리

컨트롤러에서 하던 재입력 기능을 객체에 맞게 InputView의 일로 분리

ippnsj committed:
refactor: 도메인에서의 전달값 검증 추가

ippnsj committed:
refactor: 프로퍼티 가시성 높임

가공 없이 값을 넘겨주는 getter가 필요한 dto 프로퍼티의 접근제한자를 private에서 public으로 변경

ippnsj committed:
refactor: 클래스명에서 DTO 제거

객체의 역할에 맞는 이름 설정을 위해 클래스명에서 DTO 제거

malibinYun commented on src/main/kotlin/Application.kt:
```diff
@@ -1,10 +1,12 @@
 
 fun main() {
+    val generator = RandomNumberGenerator()
     val inputView = InputView()
     val outputView = OutputView()
 
-    Controller(inputView, outputView).run()
+    Controller(generator, inputView, outputView).run()
 }
```
의존성 주입을 잘 해주셨네요! 👍👍

malibinYun commented on src/main/kotlin/controller/Controller.kt:
```diff
@@ -1,16 +1,16 @@
 ) {
-    private val cars = mutableListOf<Car>()
-    private val judgement = Judgement(cars)
+    private lateinit var cars: Cars
```
필드를 두어 값을 준비하기보다, 필요한 메서드에게 값을 만들어 넘겨주게 변경해보는 것은 어떨까요?

malibinYun replied:
이 부분 꼭 도전해보셨으면 좋겠어요 😊

malibinYun commented on src/main/kotlin/controller/Controller.kt:
```diff
@@ -61,13 +43,14 @@ class Controller(
     }
```
이 부분은 repeat()를 활용해볼 수 있겠네요 :)

malibinYun commented on src/main/kotlin/controller/Controller.kt:
```diff
@@ -61,13 +43,14 @@ class Controller(
-        cars.map { car ->
-            outputView.printRaceResult(car.race())
+        val raceOneTimeResults = cars.raceOneTime()
+        raceOneTimeResults.map { result ->
+            outputView.printRaceResult(result)
```
map은 List를 반환합니다. 다른 확장함수를 찾아볼까요? 🙂

malibinYun commented on src/main/kotlin/domain/Car.kt:
```diff
@@ -1,42 +1,56 @@
+        verifyNameLength(name)
+    }
 
-    fun race(): CarMetadataDTO {
+    fun race(): CarMetadata {
```
CarMetadata라는 이름은 다른 개발자가 보았을 때 바로 어떤 의미인지 파악할 수 있을까요?

malibinYun commented on src/main/kotlin/domain/Car.kt:
```diff
@@ -1,42 +1,56 @@
+        return CarMetadata(distance, name)
+    }
+
+    fun compareDistance(winnerDistance: Int): Int {
+        return max(winnerDistance, distance)
     }
```
메서드 명으로만 봤을 때는 어떤 동작을 하는 지 추측하기 어려워 보입니다.
외부에서 distance를 꺼내서 비교해보는 것은 어떨까요? 혹은, distance와 큰지 작은지에 대한 메서드를 만들어볼 수도 있겠네요 :)

malibinYun commented on src/main/kotlin/domain/Car.kt:
```diff
@@ -1,42 +1,56 @@
-            difference > 0 -> ComparisonResult.WIN
-            difference < 0 -> ComparisonResult.LOSE
-            else -> ComparisonResult.DRAW
+    fun isWinner(winnerDistance: Int): String {
+        if (winnerDistance == distance) {
+            return name
         }
+        return ""
+    }
```
파라미터에 어떤 값을 넣어야 하는 지 예측이 어려워보여요.
또, isWinner인데 String을 반환하는 것은 어색합니다. is~네이밍은 Boolean을 반환하는 것이 일반적이에요 :)
이것 또한 위 코멘트 처럼 distance를 꺼내 비교해보는 것은 어떨까요?

malibinYun commented on src/main/kotlin/domain/Car.kt:
```diff
@@ -1,42 +1,56 @@
+    }
+
+    private fun verifyName(name: String) {
+        name.map {
+            require(it.code in 97..122) { "$ERROR_NAME\n잘못된 자동차 이름 : $name" }
+        }
```
해당 숫자가 어떤 값을 의미하는 지 알기 어려워보입니다.
또, Char값만 비교한다면, 이후에 이름에 대한 다른 요구사항이 추가되었을 때 빠르게 대응하기는 어려워보여요.
현재는 들어와선 안되는 케이스 정도만 비교해줘도 괜찮다고 생각해요 :)
혹은 다른 방법을 고민해보셔도 좋겠어요 🙂

malibinYun commented on src/main/kotlin/domain/Cars.kt:
```diff
@@ -0,0 +1,48 @@
+import dto.Winners
+
+class Cars(generator: NumberGenerator, names: List<String>) {
+    private val cars = mutableListOf<Car>()
+
```
내부에서 list의 값을 변형하지 않기 때문에 mutableList일 필요는 없어보여요 :)

ippnsj replied:
앗! cars 리스트를 초기화할 때 add 함수를 사용하여 객체를 리스트에 넣어서 mutableList를 사용했어욤~!ㅎㅎ

malibinYun replied:
생성자에서만 호출해주고 있기 때문에, name으로 car에 대한 list를 생성하고, 바로 cars에 할당해주면 mutableList를 활용하지 않을 수 있습니다 😊

malibinYun commented on src/main/kotlin/domain/Cars.kt:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    private fun getWinners(winnerDistance: Int): Winners {
+        val winners = mutableListOf<String>()
+        cars.map { car -> winners.add(car.isWinner(winnerDistance)) }
```
map { } 의 반환 타입은 무엇일까요?

malibinYun commented on src/main/kotlin/utils/CarNamesValidator.kt:
```diff
@@ -0,0 +1,39 @@
+package utils
+
+class CarNamesValidator : Validator {
+    override fun validate(input: String) {
+        val names = input.split(',')
```
이전 피드백 처럼, 다른 검증 로직들은 필요한 객체에 위임해서 응집도를 높혀보는 것은 어떨까요? 🙂

ippnsj replied:
split은 따로 검증이 필요하지 않다고 생각해서 분리하지 않았는데 검증이 필요할까요? 🤓

malibinYun replied:
네, split은 view에서만 해주어도 됩니다 :)
validator 객체를 따로 만들어 한꺼번에 관리하기 보다, 검증이 필요한 곳에서 해당 로직을 응집도 있게 구현하시면 좋겠어요 :)

malibinYun commented on src/test/kotlin/domain/CarTest.kt:
```diff
@@ -2,22 +2,38 @@ package domain
+import org.junit.jupiter.params.provider.ValueSource
 
 class CarTest {
+    private val generator = object : NumberGenerator {
+        private val numbers = mutableListOf(2, 4, 3, 9, 5, 0)
+
+        override fun generate(): Int {
+            return numbers.removeAt(0)
+        }
+    }
```
https://github.com/woowacourse/kotlin-racingcar/pull/36#discussion_r1101785486
NumberGenerator는 필요한 곳에서 람다로 넘겨보는 것은 어떨까요?

malibinYun reviewed (changes requested):
2단계 미션 고생 많으셨습니다 :)
생각해보면 좋을 점들에 코멘트 남겨두었어요! 😃
질문 주신 점은 아래 코멘트 남겨둘게요 :)
피드백 반영하시고 다시 리뷰 요청 부탁드려요!

malibinYun commented:
> Controller의 cars 멤버 변수의 경우, 재활용 가능한 방법이 뭘까 고민을 많이 하다가 Cars 라는 일급 컬렉션을 만들어 보는건 어떨까 생각했습니다!

일급 컬렉션은 지금 내용이 아니었지만 열심히 공부해주신 것 같아요 👍
멤버변수 활용은 멀티스레드 환경에서 취약해질 수 있답니다. 먼저, 필요한 곳에 넘겨주는 형태로 연습해보아요 :)

> 입력을 다시 받는 것은 InputView의 일로 분리해 보라는 피드백을 받고 검증 위치에 대해 많이 고민했습니다. 슬랙에 질문을 올리고 여러 크루들로부터 값을 직접 사용하는 도메인에서 검증을 하는 것이 좋지 않을까요 라는 의견을 받고 도메인에서의 검증을 생각하고 있었기 때문에 고민이 컸던 것 같아요! 하지만 InputView에서 입력을 다시 받아야 하는지 여부를 결정하기 위해서는 검증이 필요하다고 생각했고, 저 또한 입력을 다시 받는 것은 컨트롤러가 아닌 InputView의 역할이 맞다고 생각했습니다!
결과적으로 도메인에서 값을 전달 받은 직후에도, InputView가 입력을 받은 직후에도 검증이 필요하다고 생각하여 두 곳에서 모두 검증을 진행하였습니다~!😄

InputView에서 최소한으로 검증해주시면 되어요.
예를 들어서, Integer를 입력 받아야하는데 integer가 아닌 경우만 걸러내면 됩니다. 이게 시도 횟수라서 1~20을 검증하는 것은 domain 로직이기에, 해당 검증을 해주어야하는 곳에서 수행하면 되어요 :)

> 남겨주신 프로퍼티 가시성 문서를 보고 주생성자에서 생성된 프로퍼티의 getter를 어떻게 가시성 높게 표현할 수 있을까 고민했습니다. 그러다가 그냥 접근제한자를 public으로 바꾸면 좋지 않을까 라는 생각을 하게 되었고 그렇게 수정했습니다! 더 좋은 방법이 있다면 알려주세요~!😊

잘 해주셨다고 생각해요! 👍

> Inner class 였던 TestNumberGenerator를 테스트 코드에 익명 클래스로 넘겨주었습니다! 람다로 넘기라는 말씀에 부합하는지는 잘 모르겠습니다😭

이 부분은 리뷰에 남겨두었어요! `fun interface`에 대해 더 찾아보시면 힌트가 되실거예요 :)

> 그동안은 하드 코딩은 무조건 나쁘다고만 생각해 보았지 하드 코딩이 더 읽기 편할수도 있다는 생각을 해보지 못 했던 것 같아요! 피드백을 받고 테스트 코드를 하드 코딩 했을 때와 안 했을 때를 비교해 보았을 때, 오히려 하드 코딩하지 않았을 때 더 복잡하다고 느꼈습니다!

생각보다 테스트에서는 일반적인 코드를 작성하는 것과 거꾸로하는 게 더 좋은 경우가 많더라구요!
중복되게, 하드코딩으로 작성하는 게 더 장점이 되기도 합니다 :)

ippnsj committed:
refactor: for 대신 고차함수인 repeat 사용

시도 횟수만큼 repeat 으로 반복

ippnsj committed:
refactor: map 대신 forEach 사용

list 반환이 불필요한 경우, map 대신 forEach 사용

ippnsj committed:
refactor: 직관적인 클래스명으로 변경

CarMetadata 라는 클래스명 대신 좀 더 직관적인 RaceResult 로 변경

ippnsj committed:
refactor: 더 큰 값인지 알려주는 메서드 추가

distance 가 이전 winnerDistance 보다 더 큰지 알려주는 메서드를 추가하여 compareDistance 메서드의 역할이 좀 더 직관적으로 보일 수 있게 수정

ippnsj committed:
refactor: isWinner 메서드명 변경

is 로 시작하는 메서드가 String 을 반환하는 것이 어색하기 때문에 isWinner 라는 Boolean 값을 반환하는 메서드를 따로 만들고 String 반환 메서드는 getWinnerName 으로 메서드명 변경

ippnsj committed:
refactor: isWinner 메서드명 변경

isWinner 라는 메서드명에 winningDistance 파라미터를 받는 것이 어색하여 isSameDistance 로 메서드명 변경

ippnsj committed:
refactor: char 코드값 비교 대신 함수 사용

char 의 코드값이 어떤 값인지 명확하지 않아 코드값 비교 대신 lower case 인지 확인해주는 함수 사용

ippnsj committed:
refactor: NumberGenerator 람다로 전달

malibinYun commented on src/main/kotlin/domain/Car.kt:
```diff
@@ -1,42 +1,66 @@
+        return RaceResult(distance, name)
+    }
+
+    fun compareDistance(winnerDistance: Int): Int {
+        if (isLongerDistance(winnerDistance)) {
+            return distance
+        }
+        return winnerDistance
+    }
+
+    fun getWinnerName(winnerDistance: Int): String {
+        if (isSameDistance(winnerDistance)) {
+            return name
+        }
+        return ""
```
winnerDistance로 들어오는 integer 값이 반드시 승자의 값만 들어올까요?
다른 개발자들이 의도에 맞지 않게 사용하게 된다면, 전혀 다른 결과를 낼 것으로 보여요.
또, 메서드의 내용을 유추하기도 쉽지 않아보입니다.

distance를 public으로 열거나, isLongerDistance를 public으로 공개하는 것도 방법이겠네요 :)

malibinYun commented on src/main/kotlin/domain/Car.kt:
```diff
@@ -1,42 +1,66 @@
-            difference < 0 -> ComparisonResult.LOSE
-            else -> ComparisonResult.DRAW
+    private fun verifyName(name: String) {
+        name.forEach {
+            require(it.isLowerCase()) { "$ERROR_NAME\n잘못된 자동차 이름 : $name" }
```
올바른 자동차 이름을 꼭 검증하고 싶으시면, Regex을 활용해보세요 :)

malibinYun commented on src/test/kotlin/domain/CarTest.kt:
```diff
@@ -2,22 +2,63 @@ package domain
+    @ValueSource(strings = ["Pobi", "woni ", " jun", "po bi", " "])
+    fun `자동차 이름에 공백이 있거나 영문 소문자가 아니면 에러가 발생한다`(input: String) {
+        val exception = assertThrows<IllegalArgumentException> {
+            Car(
+                object : NumberGenerator {
+                    private val numbers = mutableListOf(2, 4, 3, 9, 5, 0)
+
+                    override fun generate(): Int {
+                        return numbers.removeAt(0)
+                    }
+                },
```
fun interface를 공부해서 람다로 넘겨보는 것을 도전해보세요 :)

malibinYun commented on src/test/kotlin/domain/CarTest.kt:
```diff
@@ -2,22 +2,63 @@ package domain
+            },
+            name,
+        )
+        val distances = listOf(0, 1, 1, 2, 3, 3)
+        distances.forEach { distance ->
+            assertThat(car.race().distance).isEqualTo(distance)
```
테스트에 로직이 들어가면 가독성을 해칩니다.
검증부는 하드코딩해보는 것은 어떨까요?

https://github.com/woowacourse/kotlin-racingcar/pull/36#discussion_r1102715370
이때 공유드렸던 글을 다시 한 번 읽어보세요 😊

malibinYun reviewed (changes requested):
2단계 미션 피드백 반영하시느라 고생 많으셨어요!
코드가 점점 개선되어가는 모습이 보여서 좋습니다 😊

마지막으로 더 미션을 다듬어보아요 :)
특히, fun interface를 활용해서 람다로 로직을 구성해보세요!

피드백 반영해보시고 다시 리뷰 요청 부탁드릴게요!

malibinYun reviewed (approved):
리뷰 제출 시간이 지난 것을 미처 못봤었네요 😥
아쉽지만 규칙대로 이만 미션은 머지하겠습니다.
궁금한 점이 있으시면 슬랙 DM으로 물어보셔도 좋아요 😊
다른 미션들도 잘 수행하시길 응원할게요!
